File: translation/src/main/kotlin/yaku/translation/engine/onnx/OrtModel.kt
package yaku.translation.engine.onnx

import ai.onnxruntime.OnnxTensor
import ai.onnxruntime.OrtEnvironment
import ai.onnxruntime.OrtSession
import java.io.Closeable
import java.io.File
import java.nio.FloatBuffer
import java.nio.LongBuffer

/**
 * Thin wrapper over an ONNX Runtime session.
 *
 * Sessions are expensive to create (tens to hundreds of ms) and hold a large native allocation,
 * so the pipeline creates them once and keeps them alive for as long as translation is switched
 * on, then closes them when the reader exits.
 */
class OrtModel private constructor(
    private val env: OrtEnvironment,
    private val session: OrtSession,
) : Closeable {

    val inputNames: Set<String> get() = session.inputNames
    val outputNames: Set<String> get() = session.outputNames

    /**
     * Picks the first of [candidates] the graph actually exposes, falling back to the first
     * declared input. Export toolchains disagree on naming (`pixel_values` vs `input`), and a
     * pack author should not have to spell every name out in the manifest to use a stock export.
     */
    fun resolveInput(vararg candidates: String): String =
        candidates.firstOrNull { it in inputNames } ?: inputNames.first()

    fun resolveOutput(vararg candidates: String): String =
        candidates.firstOrNull { it in outputNames } ?: outputNames.first()

    fun hasInput(name: String): Boolean = name in inputNames

    fun floatTensor(data: FloatArray, shape: LongArray): OnnxTensor =
        OnnxTensor.createTensor(env, FloatBuffer.wrap(data), shape)

    fun longTensor(data: LongArray, shape: LongArray): OnnxTensor =
        OnnxTensor.createTensor(env, LongBuffer.wrap(data), shape)

    /**
     * Runs the session and hands every output to [block].
     *
     * The float path below carries one tensor, which suits a detector whose whole answer is a
     * probability map. One that returns boxes has three - labels, boxes and scores - and the
     * labels are integers, so they cannot come back through a FloatArray at all.
     */
    fun <T> run(inputs: Map<String, OnnxTensor>, block: (OrtSession.Result) -> T): T =
        session.run(inputs).use(block)

    /**
     * Runs the session and hands the named output to [block] as a flat float array plus its shape.
     *
     * The result is copied out before the native buffers are released, so [block] may keep it.
     */
    fun <T> run(
        inputs: Map<String, OnnxTensor>,
        outputName: String?,
        block: (data: FloatArray, shape: LongArray) -> T,
    ): T = session.run(inputs).use { results ->
        val tensor = results.tensor(outputName)
        val buffer = tensor.floatBuffer
        val data = FloatArray(buffer.remaining())
        buffer.get(data)
        block(data, tensor.info.shape)
    }

    /**
     * Runs one decoder step and returns the token with the highest logit at the final position.
     *
     * A decoder answers with logits for every position so far, `[1, steps, vocab]`, and greedy
     * decoding reads only the last row. Copying the whole tensor into an array first, as [run]
     * does, allocated a vocabulary-sized row for every step already taken - on every step, on top
     * of the copy ONNX Runtime makes itself.
     */
    fun argmaxAtLastStep(inputs: Map<String, OnnxTensor>, outputName: String?): Int =
        session.run(inputs).use { results ->
            val tensor = results.tensor(outputName)
            argmaxOfLastRow(tensor.floatBuffer, rowLength = tensor.info.shape.last().toInt())
        }

    private fun OrtSession.Result.tensor(outputName: String?): OnnxTensor {
        val value = if (outputName != null && outputNames.contains(outputName)) {
            get(outputName).orElseThrow { IllegalStateException("No output '$outputName'") }
        } else {
            get(0)
        }
        return value as? OnnxTensor ?: error("Output '${outputName ?: 0}' is not a tensor")
    }

    override fun close() {
        runCatching { session.close() }
    }

    companion object {
        fun open(file: File, threads: Int): OrtModel {
            require(file.exists()) { "Model file missing: ${file.name}" }
            val env = OrtEnvironment.getEnvironment()
            val options = OrtSession.SessionOptions().apply {
                setIntraOpNumThreads(threads.coerceAtLeast(1))
                setOptimizationLevel(OrtSession.SessionOptions.OptLevel.ALL_OPT)

                // The CPU arena keeps every block it has ever allocated, so a run of pages
                // never gives memory back: measured across one chapter the native heap went
                // from 141MB two minutes in to 593MB at four, and the process was killed at
                // six while still in the foreground. Pages differ in size, so the arena cannot
                // reuse much of what it holds anyway, and the memory pattern planner has
                // little to plan for. Both cost some allocation speed per page and keep the
                // footprint flat, which is what lets a chapter finish at all.
                setCPUArenaAllocator(false)
                setMemoryPatternOptimization(false)
            }
            return OrtModel(env, env.createSession(file.absolutePath, options))
        }
    }
}

/**
 * An encoder's output, copied out of its session so every decoder step can be fed the same state.
 */
internal class EncoderState(val data: FloatArray, val shape: LongArray)

/** Index of the largest of the last [rowLength] values - the final step's best token. */
internal fun argmaxOfLastRow(values: FloatBuffer, rowLength: Int): Int {
    val offset = values.limit() - rowLength
    var best = 0
    var bestValue = Float.NEGATIVE_INFINITY
    for (i in 0 until rowLength) {
        val value = values.get(offset + i)
        if (value > bestValue) {
            bestValue = value
            best = i
        }
    }
    return best
}

File: python/src/yaku_translate/ort_model.py
"""Thin wrapper over an ONNX Runtime session.

A port of `yaku.translation.engine.onnx.OrtModel`. Sessions are expensive to create (tens to
hundreds of ms) and hold a large native allocation, so the pipeline creates them once and
keeps them alive for as long as translation is switched on, then closes them at the end.
"""

from __future__ import annotations

import os
from pathlib import Path

import numpy as np
import onnxruntime as ort


def default_threads() -> int:
    """Half the machine's cores, capped - the same shape as the Android default.

    The cap is not about politeness. ONNX Runtime's intra-op pool spends more time
    synchronising than computing on the small tensors this pipeline runs.
    """
    cores = os.cpu_count() or 2
    return min(max(cores // 2, 1), 4)


class OrtModel:
    def __init__(self, session: ort.InferenceSession) -> None:
        self._session = session
        self.input_names: tuple[str, ...] = tuple(i.name for i in session.get_inputs())
        self.output_names: tuple[str, ...] = tuple(o.name for o in session.get_outputs())

    @classmethod
    def open(cls, path: Path | str, threads: int) -> OrtModel:
        path = Path(path)
        if not path.exists():
            raise FileNotFoundError(f"Model file missing: {path.name}")

        options = ort.SessionOptions()
        options.intra_op_num_threads = max(threads, 1)
        options.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_ALL

        # The CPU arena keeps every block it has ever allocated, so a run of pages never
        # gives memory back: measured on Android across one chapter the native heap went from
        # 141MB two minutes in to 593MB at four, and the process was killed at six while
        # still in the foreground. Pages differ in size, so the arena cannot reuse much of
        # what it holds anyway, and the memory pattern planner has little to plan for. Both
        # cost some allocation speed per page and keep the footprint flat, which is what lets
        # a long run finish at all.
        options.enable_cpu_mem_arena = False
        options.enable_mem_pattern = False

        return cls(ort.InferenceSession(str(path), options, providers=["CPUExecutionProvider"]))

    # -- name resolution ---------------------------------------------------------------

    def resolve_input(self, *candidates: str) -> str:
        """Picks the first of ``candidates`` the graph actually exposes, falling back to the
        first declared input.

        Export toolchains disagree on naming (`pixel_values` vs `input`), and a pack author
        should not have to spell every name out in the manifest to use a stock export.
        """
        return next((c for c in candidates if c in self.input_names), self.input_names[0])

    def resolve_output(self, *candidates: str) -> str:
        return next((c for c in candidates if c in self.output_names), self.output_names[0])

    def has_input(self, name: str) -> bool:
        return name in self.input_names

    def first_input_matching(self, *names: str) -> str | None:
        return next((n for n in names if n in self.input_names), None)

    # -- inference ---------------------------------------------------------------------

    def run_all(self, inputs: dict[str, np.ndarray]) -> dict[str, np.ndarray]:
        """Runs the session and returns every output by name.

        The float path below carries one tensor, which suits a detector whose whole answer is
        a probability map. One that returns boxes has three - labels, boxes and scores - and
        the labels are integers, so they cannot come back through a float array at all.
        """
        results = self._session.run(None, inputs)
        return dict(zip(self.output_names, results, strict=True))

    def run(self, inputs: dict[str, np.ndarray], output_name: str | None = None) -> np.ndarray:
        """Runs the session and returns the named output.

        Kotlin hands back a flat float array plus its shape because it has to copy out of the
        native buffer before it is released; numpy already owns its array, so the shape rides
        along on it.
        """
        if output_name is not None and output_name in self.output_names:
            requested = [output_name]
        else:
            requested = [self.output_names[0]]
        return self._session.run(requested, inputs)[0]

    def close(self) -> None:
        self._session = None  # type: ignore[assignment]

    def __enter__(self) -> OrtModel:
        return self

    def __exit__(self, *_exc) -> None:
        self.close()


def float_tensor(data: np.ndarray, shape: tuple[int, ...]) -> np.ndarray:
    return np.ascontiguousarray(data, dtype=np.float32).reshape(shape)


def long_tensor(data, shape: tuple[int, ...]) -> np.ndarray:
    return np.ascontiguousarray(data, dtype=np.int64).reshape(shape)
